concert/app/src/functions/aggregate_writer/index.py

```diff
@@ -1,2 +1,116 @@
+DYNAMODB_ENDPOINT = os.environ.get("DYNAMODB_ENDPOINT")
+
+dynamodb_args = {}
+if DYNAMODB_ENDPOINT:
+    dynamodb_args["endpoint_url"] = DYNAMODB_ENDPOINT
+dynamodb = boto3.resource("dynamodb", **dynamodb_args)
+table = dynamodb.Table(DYNAMODB_TABLE_NAME)
+
+CHOICE_TYPE_KEYS = {
+    CHOICE_COLOR: ATTR_CHOICE_VALUE_COLOR,
+    CHOICE_EMOTION_ENERGY: ATTR_CHOICE_VALUE_EMOTION,
+    CHOICE_EMOTION_HAPPINESS: ATTR_CHOICE_VALUE_EMOTION,
+    CHOICE_CHILLS: ATTR_CHOICE_VALUE_CHILLS,
+}
+
+
 def handler(event, context):
-    return {"statusCode": 204}
+    firehose_response = {"records": []}
+    records = event["records"]
+    for record in records:
+        record_id = record["recordId"]
+        record_response = {"recordId": record_id}
+        try:
+            put_response = add_record_to_aggregate(record)
+            record_response["result"] = "Ok"
+        except:
+            record_response["result"] = "DeliveryFailed"
+        firehose_response["records"].append(record_response)
+    return firehose_response
+
+
+def create_map_for_record_if_none_exists(timestamp, choice_key):
+    update_args = {
+        "Key": {ATTR_RECORD_ID: RECORD_ID_AGGREGATE},
+        "UpdateExpression": "SET #choice_key.#timestamp = :empty_map",
+        "ExpressionAttributeNames": {
+            "#choice_key": choice_key,
+            "#timestamp": timestamp,
+        },
+        "ExpressionAttributeValues": {":empty_map": {}},
+        "ConditionExpression": "attribute_not_exists(#choice_key.#timestamp)",
+        "ReturnValues": "NONE",
+    }
+    try:
+        return table.update_item(**update_args)
+    except ClientError as e:
+        if e.response["Error"]["Code"] != "ConditionalCheckFailedException":
+            raise
+
+
+def add_record_to_aggregate(record):
+    data = json.loads(b64decode(record["data"]).decode("utf-8"))
+    timestamp = data["CHOICE_TIME"]
+    choice_type = data["CHOICE_TYPE"]
+    choice_sum = Decimal(data["CHOICE_SUM"])
+    choice_count = data["CHOICE_COUNT"]
+
+    update_args = {
+        "Key": {ATTR_RECORD_ID: RECORD_ID_AGGREGATE},
+        "UpdateExpression": "ADD #choice_key.#timestamp.#choice_string :choice_sum, #choice_key.#timestamp.#total_choices :choice_count",
+        "ExpressionAttributeNames": {
+            "#timestamp": timestamp,
+            "#total_choices": ATTR_TOTAL_CHOICES,
+        },
+        "ExpressionAttributeValues": {
+            ":choice_sum": choice_sum,
+            ":choice_count": choice_count,
+        },
+        "ReturnValues": "NONE",
+    }
+
+    if choice_type.startswith(CHOICE_COLOR):
+        color = choice_type.split("_")[2]
+        choice_type = CHOICE_COLOR
+        update_args["ExpressionAttributeNames"]["#choice_string"] = color
+
+    elif choice_type.startswith("CHOICE_EMOTION"):
+        emotion = choice_type[7:]
+        update_args["ExpressionAttributeNames"]["#choice_string"] = emotion
+        update_args["ExpressionAttributeNames"]["#total_choices"] += "_{}".format(
```

Comment:
To make sure I understand,  is the intention here to create a record like this?
```js
{
  recordId: 'AGGREGATE',
  colors: {
    '2019-08-14T10:07:00.000Z': {
      choices: 34,
      '#AB0000': 12,
      '#00AB00': 9,
      ...
    },
    '2019-08-14T10:07:20.000Z': {
      choices: 35,
      '#AB0000': 11,
      '#00AB00': 10,
      ...
    },
    ...
  },
  emotions: {
    '2019-08-14T10:11:20.000Z': {
      choices_EMOTION_HAPPINESS: 82,
      EMOTION_HAPPINESS: 47.4,
      choices_EMOTION_ENERGY: 76,
      EMOTION_ENERGY: 21.8,
    },
    '2019-08-14T10:11:40.000Z': {
      choices_EMOTION_HAPPINESS: 87,
      EMOTION_HAPPINESS: 51.0,
      choices_EMOTION_ENERGY: 74,
      EMOTION_ENERGY: 20.2,
    },
  },
  chills: {
    '2019-08-14T10:15:40.000Z': {
      chills: 10.4,
      choices: 41,
    },
  },
}
```

I think words like "sum" and "count" and more explicit than generic "choices" for expressing the meaning of an attribute.

BTW it took me around 10 minutes of reading the code to decipher this. `#choice_string` and `#total_choices` might better be called `#choiceSum` and `#choiceCount` to help your fellow developer understand the intention. And please update the docs.

Reply:
Sorry for the confusion here; I should've included a screenshot of the output.

I've made the changes you mentioned above. Output is pictured below.

![image](https://user-images.githubusercontent.com/2954791/63150961-14c9bb80-bfbd-11e9-8dd0-d561f16cfaa1.png)